In [2]:
with open("userssharedsdfschoolimprovement2010grants.csv","r")as f:
    data=f.readlines()


In [3]:
type(data)

list

In [4]:
data[0]

'School Name,City,State,District Name,2010/11/Award Amount,Model Selected,Location\n'

In [5]:
(len(data)-1)/3

831.0

File has 3 lines of data per school. First line is header. There is data for 831 schools in the file.

In [6]:
columns=data[0]

In [7]:
columns

'School Name,City,State,District Name,2010/11/Award Amount,Model Selected,Location\n'

Pick a small portion of the data to develop codes- say 5 schools = 15 lines + 1 header line. PYTHON STARTS WITH lINE 0.

In [8]:
data[0:16]

['School Name,City,State,District Name,2010/11/Award Amount,Model Selected,Location\n',
 'HOGARTH KINGEEKUK MEMORIAL SCHOOL,SAVOONGA,AK,BERING STRAIT SCHOOL DISTRICT,$471014.00,Transformation,"200 MAIN ST\n',
 'SAVOONGA, AK 99769\n',
 '(63.6687, -170.603)"\n',
 'AKIACHAK SCHOOL,AKIACHAK,AK,YUPIIT SCHOOL DISTRICT,$520579.00,Transformation,"AKIACHAK 51100\n',
 'AKIACHAK, AK 99551\n',
 '(60.8911, -161.376)"\n',
 'GAMBELL SCHOOL,GAMBELL,AK,BERING STRAIT SCHOOL DISTRICT,$449592.00,Transformation,"169 MAIN ST\n',
 'GAMBELL, AK 99742\n',
 '(63.7413, -171.689)"\n',
 'BURCHELL HIGH SCHOOL,WASILLA,AK,MATANUSKA-SUSITNA BOROUGH SCHOOL DISTRICT,$641184.00,Transformation,"1775 WEST PARKS HWY\n',
 'WASILLA, AK 99654\n',
 '(61.5794, -149.495)"\n',
 'AKIAK SCHOOL,AKIAK,AK,YUPIIT SCHOOL DISTRICT,$399686.00,Transformation,"AKIAK 5227\n',
 'AKIAK, AK 99552\n',
 '(60.8879, -161.2)"\n']

In [9]:
data[1:16:3]

['HOGARTH KINGEEKUK MEMORIAL SCHOOL,SAVOONGA,AK,BERING STRAIT SCHOOL DISTRICT,$471014.00,Transformation,"200 MAIN ST\n',
 'AKIACHAK SCHOOL,AKIACHAK,AK,YUPIIT SCHOOL DISTRICT,$520579.00,Transformation,"AKIACHAK 51100\n',
 'GAMBELL SCHOOL,GAMBELL,AK,BERING STRAIT SCHOOL DISTRICT,$449592.00,Transformation,"169 MAIN ST\n',
 'BURCHELL HIGH SCHOOL,WASILLA,AK,MATANUSKA-SUSITNA BOROUGH SCHOOL DISTRICT,$641184.00,Transformation,"1775 WEST PARKS HWY\n',
 'AKIAK SCHOOL,AKIAK,AK,YUPIIT SCHOOL DISTRICT,$399686.00,Transformation,"AKIAK 5227\n']

By reading every third line, starting from second line,we can isolate the record for five schools in line 1 of 3 the smaller set. The number of schools can be verified with len command.

In [10]:
len (data[1:16:3])

5

In [11]:
data[2:16:3]

['SAVOONGA, AK 99769\n',
 'AKIACHAK, AK 99551\n',
 'GAMBELL, AK 99742\n',
 'WASILLA, AK 99654\n',
 'AKIAK, AK 99552\n']

By reading every third line, starting from third line,we can isolate the record for five schools in line 2 of 3 the smaller set.

In [12]:
data[3:16:3]

['(63.6687, -170.603)"\n',
 '(60.8911, -161.376)"\n',
 '(63.7413, -171.689)"\n',
 '(61.5794, -149.495)"\n',
 '(60.8879, -161.2)"\n']

By reading every third line, starting from fourth line,we can isolate the record for five schools in line 3 of 3 the smaller set. These 3 commands isolate the data in the 3 lines for each school. Now we combine the data using zip method. Also the 3 lines can be combined in to one by replacing the \n (new line) charecters with a blank space. 

In [13]:
records = []

for part1, part2, part3 in zip(data[1:16:3],data[2:16:3],data[3:16:3]):
    record=(part1+part2+part3)
    record=record.replace("\n","")
    records.append(record)


In [14]:
len(records)

5

In [15]:
records = []

for part1, part2, part3 in zip(data[1::3],data[2::3],data[3::3]):
    record=(part1+part2+part3)
    record=record.replace("\n","")
    records.append(record)

In [16]:
len(records)

831

In [17]:
for record in records[0:10]:
    print(record)


HOGARTH KINGEEKUK MEMORIAL SCHOOL,SAVOONGA,AK,BERING STRAIT SCHOOL DISTRICT,$471014.00,Transformation,"200 MAIN STSAVOONGA, AK 99769(63.6687, -170.603)"
AKIACHAK SCHOOL,AKIACHAK,AK,YUPIIT SCHOOL DISTRICT,$520579.00,Transformation,"AKIACHAK 51100AKIACHAK, AK 99551(60.8911, -161.376)"
GAMBELL SCHOOL,GAMBELL,AK,BERING STRAIT SCHOOL DISTRICT,$449592.00,Transformation,"169 MAIN STGAMBELL, AK 99742(63.7413, -171.689)"
BURCHELL HIGH SCHOOL,WASILLA,AK,MATANUSKA-SUSITNA BOROUGH SCHOOL DISTRICT,$641184.00,Transformation,"1775 WEST PARKS HWYWASILLA, AK 99654(61.5794, -149.495)"
AKIAK SCHOOL,AKIAK,AK,YUPIIT SCHOOL DISTRICT,$399686.00,Transformation,"AKIAK 5227AKIAK, AK 99552(60.8879, -161.2)"
MIDVALLEY HIGH,WASILLA,AK,MATANUSKA-SUSITNA BOROUGH SCHOOL DISTRICT,$697703.00,Restart,"7362 WEST PARKS HWY 725WASILLA, AK 99654(61.5023, -149.796)"
TULUKSAK SCHOOL,TULUKSAK,AK,YUPIIT SCHOOL DISTRICT,$427658.00,Transformation,"TULUKSAK 115TULUKSAK, AK 99679(61.1001, -160.961)"
CAPITOL HEIGHTS JR HIGH SCH,MONT

In [18]:
columns

'School Name,City,State,District Name,2010/11/Award Amount,Model Selected,Location\n'

In [19]:
columns.strip()

'School Name,City,State,District Name,2010/11/Award Amount,Model Selected,Location'

Header line was earlier defined as Columns. This line still has the new line notation, which can be eliminated using strip(). Further, comma delimiter can be used with split() method to split the 7 items in line 1 in to 7 columns.

In [20]:
columns.strip().split(",")

['School Name',
 'City',
 'State',
 'District Name',
 '2010/11/Award Amount',
 'Model Selected',
 'Location']

In [21]:
len(columns.strip().split(","))

7

Using the len() method we know that the header has 7 columns. Same logic can be applied to all records.

In [22]:
for record in records[0:10]:
    print(record.split(","))

['HOGARTH KINGEEKUK MEMORIAL SCHOOL', 'SAVOONGA', 'AK', 'BERING STRAIT SCHOOL DISTRICT', '$471014.00', 'Transformation', '"200 MAIN STSAVOONGA', ' AK 99769(63.6687', ' -170.603)"']
['AKIACHAK SCHOOL', 'AKIACHAK', 'AK', 'YUPIIT SCHOOL DISTRICT', '$520579.00', 'Transformation', '"AKIACHAK 51100AKIACHAK', ' AK 99551(60.8911', ' -161.376)"']
['GAMBELL SCHOOL', 'GAMBELL', 'AK', 'BERING STRAIT SCHOOL DISTRICT', '$449592.00', 'Transformation', '"169 MAIN STGAMBELL', ' AK 99742(63.7413', ' -171.689)"']
['BURCHELL HIGH SCHOOL', 'WASILLA', 'AK', 'MATANUSKA-SUSITNA BOROUGH SCHOOL DISTRICT', '$641184.00', 'Transformation', '"1775 WEST PARKS HWYWASILLA', ' AK 99654(61.5794', ' -149.495)"']
['AKIAK SCHOOL', 'AKIAK', 'AK', 'YUPIIT SCHOOL DISTRICT', '$399686.00', 'Transformation', '"AKIAK 5227AKIAK', ' AK 99552(60.8879', ' -161.2)"']
['MIDVALLEY HIGH', 'WASILLA', 'AK', 'MATANUSKA-SUSITNA BOROUGH SCHOOL DISTRICT', '$697703.00', 'Restart', '"7362 WEST PARKS HWY 725WASILLA', ' AK 99654(61.5023', ' -149.7

Above we did a separation of columns based on comma delimited. Upon separation we whould have 7 fields of data in the recodrs as well - corresponding to the 7 columns in the header. For the header we could check the number of fields by using len() because there is only one line of data. However in the records, we know that using len() will result in 831 items.

In [23]:
len(records)

831

To verify the number of fields in each record, we can use split on comma for record.

In [24]:
len(record.split(","))


9

we have 9 fields not seven in the sample record. We can check the entire records.

In [25]:
fields= set()
for record in records:
    fields.add(len(record.split(",")))
fields

{9, 10, 11}

The fields item above could have been defined as a list(), but list() allows duplicates, so we would see the number of fields in each of the 831 records. By defining Sets(), we prevent duplicates and see only the number of unique field lengths over the 831 records. With list () we would use append() instead of add.
This means, the recodrs have 3 different number of fields - 9, 10 and 11.

In [26]:
record

'ROOSEVELT HIGH SCHOOL,CASPER,WY,NATRONA COUNTY SCHOOL DISTRICT #1,$431382.00,Transformation,"140 EAST K STCASPER, WY 82601(42.861, -106.324)"'

In [27]:
record.split(",")

['ROOSEVELT HIGH SCHOOL',
 'CASPER',
 'WY',
 'NATRONA COUNTY SCHOOL DISTRICT #1',
 '$431382.00',
 'Transformation',
 '"140 EAST K STCASPER',
 ' WY 82601(42.861',
 ' -106.324)"']

In [28]:
Focus only on the Philadelphia School Districts.

SyntaxError: invalid syntax (<ipython-input-28-7d0fc7b48bda>, line 1)

In [29]:
for record in records:
    if "PHILADELPHIA CITY SD" in record:
        print(record)

SAYRE WILLIAM L MS,PHILADELPHIA,PA,PHILADELPHIA CITY SD,$737032.00,Transformation,"5800 WALNUT STPHILADELPHIA, PA 19139(39.9584, -75.2375)"
KING MARTIN LUTHER HS,PHILADELPHIA,PA,PHILADELPHIA CITY SD,$786964.00,Transformation,"6100 STENTON AVEPHILADELPHIA, PA 19138(40.0577, -75.1622)"
DOUGLAS STEPHEN A SCH,PHILADELPHIA,PA,PHILADELPHIA CITY SD,$319320.00,Restart,"2700 EAST HUNTINGDON STPHILADELPHIA, PA 19125(39.9767, -75.1168)"
LOCKE ALAIN SCH,PHILADELPHIA,PA,PHILADELPHIA CITY SD,$390938.00,Transformation,"4550 HAVERFORD AVEPHILADELPHIA, PA 19139(39.9625, -75.2136)"
ALCORN JAMES SCH,PHILADELPHIA,PA,PHILADELPHIA CITY SD,$338726.00,Transformation,"3200 DICKINSON STPHILADELPHIA, PA 19146(39.9354, -75.1973)"
GERMANTOWN HS,PHILADELPHIA,PA,PHILADELPHIA CITY SD,$798792.00,Transformation,"40 EAST HIGH STPHILADELPHIA, PA 19144(40.0401, -75.1769)"
ROXBOROUGH HS,PHILADELPHIA,PA,PHILADELPHIA CITY SD,$641440.00,Transformation,"6498 RIDGE AVEPHILADELPHIA, PA 19128(40.0381, -75.2228)"
UNIVERSITY CITY H

Need to check the number of fiels in the above list of schools.

In [30]:
fields= set()
for record in records:
    if "PHILADELPHIA CITY SD" in record:
        fields.add(len(record.split(",")))
fields

{9}

All the record items for Phila School Dist have 9 fields when comma separated.
We need to show School and the funding for that school. This fits well with Python Dictionary data structure. We can use that.

In [31]:
funding = {"SCHOOLNAME":[], "AMOUNT":[]}
print(funding)

{'SCHOOLNAME': [], 'AMOUNT': []}


In [32]:
funding = {"SCHOOLNAME":[], "AMOUNT":[]}
for record in records:
    if "PHILADELPHIA CITY SD" in record:
        fields=record.split(",")
        school_name=fields[0]
        amount=float(fields[4][1:])
        funding["SCHOOLNAME"].append(school_name) 
        funding["AMOUNT"].append(amount)
funding

{'SCHOOLNAME': ['SAYRE WILLIAM L MS',
  'KING MARTIN LUTHER HS',
  'DOUGLAS STEPHEN A SCH',
  'LOCKE ALAIN SCH',
  'ALCORN JAMES SCH',
  'GERMANTOWN HS',
  'ROXBOROUGH HS',
  'UNIVERSITY CITY HS',
  'SMEDLEY FRANKLIN SCH',
  'WEST PHILADELPHIA HS',
  'HARRITY WILLIAM F SCH',
  'FRANKFORD HS',
  'VAUX ROBERTS HS',
  'CLEMENTE ROBERTO MS',
  'FELS SAMUEL HS',
  'ALLEN DR ETHEL SCH',
  'FELTONVILLE INTERMEDIATE SCHOOL',
  'SOUTH PHILADELPHIA HS',
  'OLNEY HS WEST',
  'EDISON HS/FAREIRA SKILLS',
  'STETSON JOHN B MS',
  'DAROFF SAMUEL SCH',
  'BLUFORD GUION EL SCH',
  'CLYMER GEORGE SCH',
  'MANN WILLIAM B SCH',
  'GRATZ SIMON HS',
  'LINCOLN HS'],
 'AMOUNT': [737032.0,
  786964.0,
  319320.0,
  390938.0,
  338726.0,
  798792.0,
  641440.0,
  923999.0,
  362773.0,
  671142.0,
  980283.0,
  792001.0,
  466665.0,
  333327.0,
  974132.0,
  247020.0,
  244855.0,
  579958.0,
  272245.0,
  1.0,
  888708.0,
  658247.0,
  907129.0,
  392068.0,
  794021.0,
  508486.0,
  14.0]}

The amount shown here was not a number, but a string. Need to convert to numbers. We can select the amount items read everything after the first charecter ($) and define that as a float.

In [33]:
Mean_Funding= sum(funding['AMOUNT'])/len(funding['AMOUNT'])
print("Mean Funding Amount = ",Mean_Funding)

Mean Funding Amount =  555936.5185185185


In [34]:
Number_of_Schools=len(funding["SCHOOLNAME"])
print("NUMBER OF SCHOOLS = ",Number_of_Schools)

NUMBER OF SCHOOLS =  27


In [35]:
Min_Funding= min(funding['AMOUNT'])
print("Minimum Funding Amount = ",Min_Funding)

Minimum Funding Amount =  1.0


In [36]:
Max_Funding= max(funding['AMOUNT'])
print("Maximum Funding Amount = ",Max_Funding)

Maximum Funding Amount =  980283.0


In [37]:
amounts=funding["AMOUNT"]
print(amounts)

[737032.0, 786964.0, 319320.0, 390938.0, 338726.0, 798792.0, 641440.0, 923999.0, 362773.0, 671142.0, 980283.0, 792001.0, 466665.0, 333327.0, 974132.0, 247020.0, 244855.0, 579958.0, 272245.0, 1.0, 888708.0, 658247.0, 907129.0, 392068.0, 794021.0, 508486.0, 14.0]


In [38]:
print(Mean_Funding)

555936.5185185185


In [39]:
len(amounts)

27

In [40]:
def variance (item):
       return (sum([(x - Mean_Funding) ** 2 for x in item]))
var = variance (funding['AMOUNT'])
print(var)


2158816173644.7407


In [41]:
SD= (var/(len(amounts)))**0.5
print("Standard Deviation= ",SD)

Standard Deviation=  282765.19336621836


In [45]:
for x in amounts:
    print((x-Mean_Funding)**2)

32795573413.009594
53373697199.67626
55987376835.824425
27224511113.305904
47180409355.08369
58978784885.60219
7310845345.454043
135469990274.30588
37312144886.454056
13272302963.379967
180069936345.7133
55726439417.120705
7969404018.602199
49554997735.046646
174887460731.5281
95429415413.6022
96771711163.7874
577031572.5651567
80480877679.34294
309064300750.45404
110736858887.37996
10467434620.97256
123336159049.1207
26852891361.454052
56684220322.305885
2251551707.676271
309049846595.9726
